addons/website_slides/models/slide_channel.py

```diff
@@ -104,24 +105,21 @@ def _set_as_completed(self, completed=True):
             True if we make the slide as completed
             False if we remove user completion
         """
-        partner_karma = dict.fromkeys(self.mapped('partner_id').ids, 0)
-        for record in self:
-            record.completed = completed
-            partner_karma[record.partner_id.id] += record.channel_id.karma_gen_channel_finish
-
-        partner_karma = {
-            partner_id: karma_to_add
-            for partner_id, karma_to_add in partner_karma.items() if karma_to_add > 0
-        }
-
-        if partner_karma:
-            users = self.env['res.users'].sudo().search([('partner_id', 'in', list(partner_karma.keys()))])
-            for user in users:
-                karma = partner_karma[user.partner_id.id]
-                if not completed:
-                    # Mark the channel as not-completed, we remove the gained karma
-                    karma *= -1
-                users.add_karma(karma)
+        for channel, partners in groupby(self, lambda partner: partner.channel_id):
+            partners = reduce(lambda a, b: a | b, partners)
+            partners.completed = completed
+            karma = channel.karma_gen_channel_finish
+            if karma > 0:
+                users = self.env['res.users'].sudo().search([('partner_id', 'in', partners.partner_id.ids)])
+                karma_per_users = {}
+                for user in users:
+                    karma_per_users[user] = {
+                        'gain': karma if completed else karma * -1,
+                        'source': channel,
+                        'reason': _('Course Finished') if completed else _('Course Set Uncompleted'),
+                    }
+
+                self.env['res.users']._add_karma_batch(karma_per_users)
```

Comment:
I can't answer to my previous comment so I answer here:

I think it's better like this, even though the numbers were correct, you would have seen "4500 karma gained because joined course X" when that specific course would have only 1500 karma gained upon entry.
(Because you joined another course at the same time that made you gain 3000 karma).
This version is much more explicit.

(In the end everything will be consolidated but hey, this task is about having a nice tracking, so let us have one).